File: bulk-merge-prs.js
#!/usr/bin/env node

import { execSync } from 'child_process';
import fs from 'fs';

const GITHUB_TOKEN = 'ghs_9MCqZ1UVpKmpFUYQWtQ92vycUmAt7Y21ivzT';
const REPO = 'Zion-Holdings/zion.app';
const API_BASE = `https://api.github.com/repos/${REPO}`;

async function fetchOpenPRs() {
  try {
    const response = await fetch(`${API_BASE}/pulls?state=open&per_page=100`, {
      headers: {
        'Authorization': `token ${GITHUB_TOKEN}`,
        'Accept': 'application/vnd.github.v3+json'
      }
    });
    
    if (!response.ok) {
      throw new Error(`GitHub API error: ${response.status}`);
    }
    
    return await response.json();
  } catch (error) {
    console.error('Error fetching PRs:', error.message);
    return [];
  }
}

async function mergePR(prNumber, title, headRef) {
  console.log(`\n🔄 Processing PR #${prNumber}: ${title}`);
  console.log(`   Branch: ${headRef}`);
  
  try {
    // Fetch the branch
    console.log('   📥 Fetching branch...');
    execSync(`git fetch origin ${headRef}`, { stdio: 'pipe' });
    
    // Try to merge
    console.log('   🔀 Attempting merge...');
    try {
      execSync(`git merge origin/${headRef} --no-ff -m "Merge PR #${prNumber}: ${title}"`, { stdio: 'pipe' });
      console.log(`   ✅ Successfully merged PR #${prNumber}`);
      return true;
    } catch (mergeError) {
      console.log('   ⚠️  Merge conflicts detected, attempting resolution...');
      
      // Try to resolve conflicts automatically
      try {
        // Use git checkout --theirs for most conflicts
        execSync(`git status --porcelain | grep "^UU" | cut -c4- | xargs -I {} git checkout --theirs {}`, { stdio: 'pipe' });
        
        // Add resolved files
        execSync(`git add .`, { stdio: 'pipe' });
        
        // Commit the resolution
        execSync(`git commit -m "Resolve conflicts for PR #${prNumber}: ${title}"`, { stdio: 'pipe' });
        
        console.log(`   ✅ Resolved conflicts and merged PR #${prNumber}`);
        return true;
      } catch (resolveError) {
        console.log(`   ❌ Could not resolve conflicts for PR #${prNumber}`);
        // Reset the merge
        execSync(`git merge --abort`, { stdio: 'pipe' });
        return false;
      }
    }
  } catch (error) {
    console.log(`   ❌ Error processing PR #${prNumber}: ${error.message}`);
    return false;
  }
}

async function main() {
  console.log('🚀 Starting bulk PR merge process...\n');
  
  // Ensure we're on main branch
  try {
    execSync('git checkout main', { stdio: 'pipe' });
    console.log('✅ Switched to main branch');
  } catch (error) {
    console.error('❌ Failed to switch to main branch:', error.message);
    return;
  }
  
  // Fetch latest changes
  try {
    execSync('git pull origin main --no-rebase', { stdio: 'pipe' });
    console.log('✅ Pulled latest changes from main');
  } catch (error) {
    console.log('⚠️  Pull had conflicts, continuing...');
  }
  
  // Get open PRs
  console.log('📋 Fetching open PRs...');
  const prs = await fetchOpenPRs();
  
  if (prs.length === 0) {
    console.log('✅ No open PRs found');
    return;
  }
  
  console.log(`📊 Found ${prs.length} open PRs`);
  
  let successCount = 0;
  let failCount = 0;
  
  // Process each PR
  for (const pr of prs) {
    const success = await mergePR(pr.number, pr.title, pr.head.ref);
    if (success) {
      successCount++;
    } else {
      failCount++;
    }
    
    // Small delay to avoid rate limiting
    await new Promise(resolve => setTimeout(resolve, 1000));
  }
  
  console.log(`\n📊 Merge Summary:`);
  console.log(`   ✅ Successfully merged: ${successCount}`);
  console.log(`   ❌ Failed to merge: ${failCount}`);
  
  // Push all changes
  if (successCount > 0) {
    console.log('\n🚀 Pushing changes to main...');
    try {
      execSync('git push origin main --force-with-lease', { stdio: 'pipe' });
      console.log('✅ Successfully pushed all changes');
    } catch (error) {
      console.log('⚠️  Push failed, but merges were successful locally');
    }
  }
  
  console.log('\n🎉 Bulk PR merge process completed!');
}

// Run the script
main().catch(console.error);

File: bulk_merge_prs.py
#!/usr/bin/env python3

import requests
import json
import time
import sys

# GitHub API configuration
GITHUB_API_BASE = "https://api.github.com/repos/Zion-Holdings/zion.app"
HEADERS = {
    "Accept": "application/vnd.github.v3+json",
    "User-Agent": "PR-Merge-Script"
}

def get_open_prs():
    """Get all open pull requests"""
    url = f"{GITHUB_API_BASE}/pulls"
    params = {"state": "open", "per_page": 100}
    
    response = requests.get(url, headers=HEADERS, params=params)
    if response.status_code == 200:
        return response.json()
    else:
        print(f"Error fetching PRs: {response.status_code}")
        return []

def merge_pr(pr_number, title):
    """Merge a pull request using GitHub API"""
    url = f"{GITHUB_API_BASE}/pulls/{pr_number}/merge"
    
    data = {
        "commit_title": f"Merge PR #{pr_number}: {title}",
        "commit_message": f"Automated merge of PR #{pr_number}",
        "merge_method": "merge"
    }
    
    response = requests.put(url, headers=HEADERS, json=data)
    
    if response.status_code == 200:
        print(f"✅ Successfully merged PR #{pr_number}: {title}")
        return True
    elif response.status_code == 405:
        print(f"❌ PR #{pr_number} cannot be merged (not mergeable): {title}")
        return False
    elif response.status_code == 409:
        print(f"⚠️  PR #{pr_number} has conflicts: {title}")
        return False
    else:
        print(f"❌ Error merging PR #{pr_number}: {response.status_code} - {response.text}")
        return False

def main():
    print("Starting bulk PR merge process...")
    
    # Get all open PRs
    prs = get_open_prs()
    
    if not prs:
        print("No open PRs found.")
        return
    
    print(f"Found {len(prs)} open PRs")
    
    # Process each PR
    successful_merges = 0
    failed_merges = 0
    
    for pr in prs:
        pr_number = pr["number"]
        title = pr["title"]
        state = pr["state"]
        mergeable = pr.get("mergeable")
        
        print(f"\nProcessing PR #{pr_number}: {title}")
        print(f"State: {state}, Mergeable: {mergeable}")
        
        if state == "open" and mergeable is not False:
            if merge_pr(pr_number, title):
                successful_merges += 1
            else:
                failed_merges += 1
        else:
            print(f"⏭️  Skipping PR #{pr_number} (not mergeable)")
            failed_merges += 1
        
        # Small delay to avoid rate limiting
        time.sleep(1)
    
    print(f"\n📊 Merge Summary:")
    print(f"✅ Successfully merged: {successful_merges}")
    print(f"❌ Failed to merge: {failed_merges}")
    print(f"📝 Total processed: {len(prs)}")

if __name__ == "__main__":
    main()